lib/IDE/Refactoring.cpp

```diff
@@ -3887,7 +3889,1442 @@ bool RefactoringActionConvertToComputedProperty::performChange() {
+  const AsyncHandlerDesc &TopHandler;
+  SmallString<0> Buffer;
+  llvm::raw_svector_ostream OS;
+
+  llvm::DenseSet<const Decl *> Unwraps;
+  llvm::DenseSet<const Decl *> Placeholders;
+  llvm::DenseMap<const Decl *, std::string> Names;
+
+  SourceLoc LastAddedLoc;
+  int NestedExprCount = 0;
+
+public:
+  AsyncConversionStringBuilder(SourceManager &SM, DiagnosticEngine &DiagEngine,
+                               const AsyncHandlerDesc &TopHandler)
+      : SM(SM), DiagEngine(DiagEngine), TopHandler(TopHandler), Buffer(),
+        OS(Buffer) {
+    Placeholders.insert(TopHandler.Handler);
+  }
+
+  void replace(ASTNode Node, SourceEditConsumer &EditConsumer) {
+    CharSourceRange Range =
+        Lexer::getCharSourceRangeFromSourceRange(SM, Node.getSourceRange());
+    EditConsumer.accept(SM, Range, Buffer.str());
+    Buffer.clear();
+  }
+
+  void insertAfter(ASTNode Node, SourceEditConsumer &EditConsumer) {
+    EditConsumer.insertAfter(SM, Node.getEndLoc(), "\n\n");
+    EditConsumer.insertAfter(SM, Node.getEndLoc(), Buffer.str());
+    Buffer.clear();
+  }
+
+  void convertFunction(const FuncDecl *FD) {
+    addFuncDecl(FD);
+    convertNode(FD->getBody());
+  }
+
+  void convertNodes(ArrayRef<ASTNode> Nodes) {
+    for (auto Node : Nodes) {
+      OS << "\n";
+      convertNode(Node);
+    }
+  }
+
+  void convertNode(ASTNode Node, SourceLoc StartOverride = {},
+                   bool ConvertCalls = true) {
+    if (!StartOverride.isValid())
+      StartOverride = Node.getStartLoc();
+
+    llvm::SaveAndRestore<SourceLoc> RestoreLoc(LastAddedLoc, StartOverride);
+    llvm::SaveAndRestore<int> RestoreCount(NestedExprCount,
+                                           ConvertCalls ? 0 : 1);
+    walk(Node);
+    addRange(LastAddedLoc, Node.getEndLoc(), /*ToEndOfToken=*/true);
+  }
+
+private:
+  bool walkToDeclPre(Decl *D, CharSourceRange Range) override { return false; }
+
+#define PLACEHOLDER_START "<#"
+#define PLACEHOLDER_END "#>"
+  bool walkToExprPre(Expr *E) override {
+    NestedExprCount++;
+
+    // TODO: Handle Result.get as well
+    if (auto *DRE = dyn_cast<DeclRefExpr>(E)) {
+      if (auto *D = DRE->getDecl()) {
+        bool AddPlaceholder = Placeholders.count(D);
+        StringRef Name = newNameFor(D, false);
+        if (AddPlaceholder || !Name.empty())
+          return addCustom(DRE->getStartLoc(),
+                           Lexer::getLocForEndOfToken(SM, DRE->getEndLoc()),
+                           [&]() {
+                             if (AddPlaceholder)
+                               OS << PLACEHOLDER_START;
+                             if (!Name.empty())
+                               OS << Name;
+                             else
+                               D->getName().print(OS);
+                             if (AddPlaceholder)
+                               OS << PLACEHOLDER_END;
+                           });
+      }
+    } else if (auto *FTE = dyn_cast<ForceValueExpr>(E)) {
```

Comment:
Should we handle BindOptionalExpr too? e.g. `oldOptD?.someMember?.otherMember` -> `newNonOptD.someMember.otherMember` (or possibly `newNonOptD.someMember?.otherMember` if `someMember` is itself optional).

Reply:
Will handle later